File: backend/routers/auditoria.py
"""
Router del audit log (T18.4a) — lectura paginada de eventos de auditoría.
Sección: "auditoria". Solo lectura: el sistema escribe el audit, no el usuario.
empresa_id para el filtro: header X-Empresa-Id (get_empresa_id), None = consolidado.
"""
from datetime import date
from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Request

from schemas.auditoria import AuditLogListResponse
from services.audit_service import AuditService
from utils.empresa import get_empresa_id
from utils.permisos import Accion, Seccion, require_permission

router = APIRouter()

SECCION = Seccion.AUDITORIA


def _svc() -> AuditService:
    return AuditService()


@router.get("", response_model=AuditLogListResponse, dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def list_auditoria(
    request: Request,
    usuario_id: Optional[UUID] = Query(None),
    entidad: Optional[str] = Query(None),
    evento: Optional[str] = Query(None),
    fecha_desde: Optional[date] = Query(None),
    fecha_hasta: Optional[date] = Query(None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    service: AuditService = Depends(_svc),
) -> AuditLogListResponse:
    return service.listar(
        empresa_id=get_empresa_id(request),
        usuario_id=usuario_id,
        entidad=entidad,
        evento=evento,
        fecha_desde=fecha_desde,
        fecha_hasta=fecha_hasta,
        page=page,
        page_size=page_size,
    )

File: frontend/services/auditoria.ts
import { apiFetch } from "@/services/api"
import type { AuditLogListResponse } from "@/types/auditoria"

export interface AuditoriaFiltros {
  usuario_id?: string
  entidad?: string
  evento?: string
  registro_id?: string
  fecha_desde?: string
  fecha_hasta?: string
  page?: number
  page_size?: number
}

/**
 * Lista eventos de auditoría paginados y filtrados.
 * empresa_id NO va como query param: apiFetch ya inyecta X-Empresa-Id del empresaStore
 * (consolidado o empresa activa), igual que todos los listados.
 */
export async function fetchAuditoria(filtros: AuditoriaFiltros = {}): Promise<AuditLogListResponse> {
  const params = new URLSearchParams()
  if (filtros.usuario_id) params.set("usuario_id", filtros.usuario_id)
  if (filtros.entidad) params.set("entidad", filtros.entidad)
  if (filtros.evento) params.set("evento", filtros.evento)
  if (filtros.registro_id) params.set("registro_id", filtros.registro_id)
  if (filtros.fecha_desde) params.set("fecha_desde", filtros.fecha_desde)
  if (filtros.fecha_hasta) params.set("fecha_hasta", filtros.fecha_hasta)
  if (filtros.page) params.set("page", String(filtros.page))
  if (filtros.page_size) params.set("page_size", String(filtros.page_size))
  const query = params.size ? `?${params}` : ""
  return apiFetch<AuditLogListResponse>(`/api/auditoria${query}`)
}
